### PR TITLE
[Attn] Add Wall attention (per-channel multiplicative-decay) op, layer, and model

### PR DESCRIPTION
**Pull request by tea-more**
## Summary
Wall Attention is an attention variant with a **per-channel, per-timestep multiplicative decay** baked into the QK inner product. Where standard attention scores a pair $(i, j)$ with $\sum_n q_{i,n}\, k_{j,n}$, Wall Attention weights each channel $n$ by a learned decay accumulated between the two positions. This gives each query channel an independent, content-dependent forgetting rate, generalizing scalar gating (FoX) and RoPE-style decays to the full channel dimension. Setting $g = 0$ recovers vanilla softmax attention.

See our blog for more information: 
https://blog.tilderesearch.com/blog/wall-attn  

This PR introduces a Wall Attention kernels for training and decoding and the Wall Transformer. More concretely, the following are introduced:
1. **Reference implementation** of Wall Attention implemented in Torch, for pedagogical and testing purposes. 
2. **Wall training kernel** implemented in Triton for training (fwd+bwd) end-to-end, with GQA and sliding window support¹.
  gate and attention sink.¹
3. **Wall decode kernel** implemented in Triton for single-step decode (fwd only) plus the pre-rescaled KV-cache builder¹.
4. **A `WallAttention` layer** mirroring `ForgettingAttention` with per-channel log-decay gate via `g_proj` + `logsigmoid`. Supports the training/prefill route with the training kernel and cached decode route with the decode kernel.
5. **A `WallTransformerModel`(`ForCausalLM`) model** mirroring `ForgettingTransformer`, auto registered with the Transformers library.

*¹ The Triton kernels are currently optimized for H100 GPUs.*

## Test Plan
We ship this PR with tests for both the kernels and the modeling code:
* `tests/ops/test_wall_attn.py` tests forward / backward (dq, dk, dv) parity vs. the native Torch reference, perofrms gate-gradient checks, and cached-decode self-consistency.
* `tests/models/test_modeling_wall_transformer.py` smoke tests `WallTransformer` modeling code forward / backward.

> [!NOTE]
> Numerical tolerances are calibrated to the current torch 2.10 / Triton stack, which shows a systematic kernel-vs-reference gap larger than on FLA's pinned stack. The per-channel `dL/dg` finite-difference check is marked `xfail` for this reason.
> For the same reason, the `WallTransformerConfig` is listed in `MODELING_UNSUPPORTED_VARLEN` (as `ForgettingTransformer` is) and `GENERATION_UNSUPPORTED` although the decode path is wired and runs; it just isn't asserted at those tolerances here.

## Breaking changes
None. All additions so no existing APIs changed.

### CODE REVIEW

**Review by sustcsonglin — APPROVED** [review]

File: fla/layers/wall_attn.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+        k_tilde, r_cache = build_wall_kv_cache(k, p, WALL_DECODE_CHUNK)
+        p_curr = p[:, -q.shape[1]:].contiguous()
+        gs_cumsum = chunk_global_cumsum(gs, scale=RCP_LN2) if gs is not None else None
+        return parallel_wall_attn_decode(
+            q, v, p_curr, k_tilde, r_cache,
+            sink_bias=None,
+            scale=scale,
+            cache_chunk_size=WALL_DECODE_CHUNK,
+            g_scalar_cumsum=gs_cumsum,
+        )
```
[the remark this answers is not in the document]
Comment (tea-more) [reply fla/layers/wall_attn.py]: Rewrote the non-windowed decode to cache the pre-rescaled state and extend it evert decode step.

File: fla/models/wall_transformer/modeling_wall_transformer.py [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+            use_cache=use_cache,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+            **kwargs,
+        )
+
+        hidden_states = outputs[0]
+
+        logits = None if self.config.fuse_linear_cross_entropy else self.lm_head(hidden_states[:, -logits_to_keep:])
```
[the remark this answers is not in the document]
Comment (tea-more) [reply fla/models/wall_transformer/modeling_wall_transformer.py]: Applied guards

File: fla/models/wall_transformer/modeling_wall_transformer.py [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+                with torch.no_grad():
+                    p /= math.sqrt(num_residuals_per_layer * self.config.num_hidden_layers)
+
+
+class WallTransformerModel(WallTransformerPreTrainedModel):
+
+    def __init__(
+        self,
+        config: WallTransformerConfig,
+    ) -> WallTransformerModel:
```
[the remark this answers is not in the document]
Comment (tea-more) [reply fla/models/wall_transformer/modeling_wall_transformer.py]: Fixed

File: fla/ops/wall_attn/decode.py [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+    if p_curr.shape != (B, T_q, HQ, K):
+        raise ValueError(f"p_curr shape {p_curr.shape} != expected {(B, T_q, HQ, K)}")
+    if T_kv > NC * C:
+        raise ValueError(f"NC*C ({NC * C}) < T_kv ({T_kv}); cache anchors do not cover cache")
+    if HQ % H != 0:
+        raise ValueError(f"HQ ({HQ}) must be divisible by H ({H})")
+
+    for t in (q, v, p_curr, k_tilde, r_cache):
+        if t.stride(-1) != 1:
+            raise ValueError("decode tensors must be contiguous in the last (K or V) dim")
```
[the remark this answers is not in the document]
Comment (tea-more) [reply fla/ops/wall_attn/decode.py:205]: Added the dim stride check

File: fla/ops/wall_attn/parallel.py [hunk 5]
```diff
@@ -0,0 +1,1003 @@
...
+            Cumulative seqlens of shape ``[N + 1]`` for varlen packing
+            (requires ``B == 1``). Default: `None`.
+
+    Returns:
+        Attention output of shape ``[B, T, HQ, V]``.
+    """
+    if scale is None:
+        scale = k.shape[-1] ** -0.5
+    if cu_seqlens is not None and q.shape[0] != 1:
+        raise ValueError("`cu_seqlens` (varlen) requires batch size 1")
```
[the remark this answers is not in the document]
Comment (tea-more) [reply fla/ops/wall_attn/parallel.py:1002]: Added check

**Comment by tea-more** [comment]
@sustcsonglin Comments were addressed and the PR is ready to be merged

**Review by zhiyuan1i — APPROVED** [review]
